libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance/ExecutableSpecRule.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Test.Cardano.Ledger.Conformance.ExecutableSpecRule (
+  ExecutableSpecRule (..),
+  conformsToImpl,
+  computationResultToEither,
+) where
+
+import Cardano.Ledger.BaseTypes (Inject, ShelleyBase)
+import Cardano.Ledger.Core (EraRule)
+import qualified Constrained as CV2
+import Constrained.Base (Spec (..))
+import Control.State.Transition.Extended (STS (..))
+import Data.Bifunctor (Bifunctor (..))
+import Data.Bitraversable (bimapM)
+import Data.Typeable (Typeable)
+import GHC.Base (NonEmpty, Symbol)
+import GHC.TypeLits (KnownSymbol)
+import qualified Lib as Agda
+import Test.Cardano.Ledger.Conformance.SpecTranslate (SpecTranslate (..), runSpecTransM)
+import Test.Cardano.Ledger.Imp.Common (
+  MonadGen (..),
+  NFData,
+  ToExpr,
+  expectRightExpr,
+  shouldBeExpr,
+ )
+import Test.Cardano.Ledger.Shelley.ImpTest (
+  ImpTestM,
+  impAnn,
+  logEntry,
+  logToExpr,
+  tryRunImpRule,
+ )
+
+class
+  ( CV2.HasSpec fn (Environment (EraRule rule era))
+  , CV2.HasSpec fn (State (EraRule rule era))
+  , CV2.HasSpec fn (Signal (EraRule rule era))
+  , CV2.HasSpec fn (ExecContext fn rule era)
+  , Inject (ExecContext fn rule era) (SpecTransContext (Environment (EraRule rule era)))
+  , Inject (ExecContext fn rule era) (SpecTransContext (State (EraRule rule era)))
+  , Inject (ExecContext fn rule era) (SpecTransContext (Signal (EraRule rule era)))
+  , Inject (ExecContext fn rule era) (SpecTransContext (PredicateFailure (EraRule rule era)))
```

Comment:
I would consider abstracting over the whole `C` has to hold for `Environment (EraRule rule era)`, `State`, `Signal`, and `PredicateFailure`. That way you'd have one third the number of constraints here.